Improve error on 'async static' in class

erlliam opened the pull request:
Resolves #291.

erlliam commented:
Ignore my request for review.

erlliam commented:
It should be good to review now.

Locally the string comparison works just fine, I'm not sure what the issue is (I thought it was lack of `_sv`).

My current solution is sort of hacky... I'm sure you'd prefer the `source_code_span` to point to `async static` and not the identifier. Besides that, this solution gets the job done without having to store extra information or state.

erlliam commented:
`property_name->normalized_name() == u8"static"`
I realize this will not work. It will incorrectly assume `\u0073tatic` a keyword...

erlliam commented:
The third attempt is still silly. What I'm thinking:

* See async
* Store the start of the token
* Skip
* See static
* Set a boolean that says potentially_async_static
* Store the end of the token
* Once we are inside `parse_and_visit_field_or_method_impl`, if static isn't a method, report the error

The main issue I'd deal with the lambda captures. Should I just add another parameter to the lambdas? Perhaps a struct? No clue.

strager commented on src/quick-lint-js/parse.h:
```diff
@@ -1549,6 +1555,9 @@ class parser {
         this->skip();
       }
+      if (this->peek().type == token_type::kw_static) {
+        async_static = true;
```
I'd report the error here, and highlight the `async static` keywords instead of the method name. I think that'd simplify the logic and also be a better user experience.

erlliam replied:
That's what I tried originally, but then it triggers a false positive for an async method named static. (`async static() {}`)

Perhaps I should use the transaction stuff and just check for a left paren right then and there?

strager replied:
> That's what I tried originally, but then it triggers a false positive for an async method named static. (`async static() {}`)

I realized that soon after I made my comment. =X

strager commented on src/quick-lint-js/error.h:
```diff
@@ -1139,6 +1139,14 @@
              continue_statement))                                              \
+                                                                               \
+  QLJS_ERROR_TYPE(                                                             \
+      error_async_static_method, "E202",                                       \
```
Nit: Sort error types alphabetically by struct name.

strager commented on src/quick-lint-js/parse.h:
```diff
@@ -1482,6 +1484,10 @@ class parser {
           //   method() {}
           // }
           v.visit_property_declaration(property_name);
+        } else if (async_static) {
+          this->error_reporter_->report(error_async_static_method{
```
Document with a comment:
```suggestion
        } else if (async_static) {
          // async static method() {}             // Invalid.
          // async static *myAsyncGenerator() {}  // Invalid.
          this->error_reporter_->report(error_async_static_method{
```

strager commented on src/quick-lint-js/parse.h:
```diff
@@ -1549,6 +1555,9 @@ class parser {
         method_attributes = function_attributes::async_generator;
         this->skip();
       }
+      if (this->peek().type == token_type::kw_static) {
+        async_static = true;
```
Document with a comment:
```suggestion
      if (this->peek().type == token_type::kw_static) {
        // async static method() {}  // Invalid.
        // async static() {}
        async_static = true;
```

strager commented on test/test-parse-class.cpp:
```diff
@@ -212,6 +212,18 @@ TEST(test_parse, class_statement_with_methods) {
                 ElementsAre(spy_visitor::visited_property_declaration{u8"m"}));
   }
 
+  {
```
I'd put this test into its own `TEST`. All other tests in this `TEST` assert no error.

strager commented on test/test-parse-class.cpp:
```diff
@@ -1078,6 +1078,20 @@ TEST(test_parse, class_statement_as_with_statement_body_is_disallowed) {
+TEST(test_parse, async_static_method_is_disallowed) {
+  {
+    spy_visitor v;
+    padded_string code(u8"class C { async static m() {} }"_sv);
```
Verify that error recovery works by checking visits. In particular, a class property called `m` should be declared, and its body should be visited, and `await` should be treated as an operator inside the body.

strager reviewed (commented):
Looks good, but I'm skeptical that it's parsing correctly. I'd expect lines 1458-1460 (or similar) to execute. Adding to the tests should give me confidence.

erlliam commented:
Actually, this won't work. What I would like to do is use goto next (the original next, not the one I added), but I don't know how to do that. I guess the original next doesn't exist inside of the lambda. I'm fighting C++.

strager reviewed (commented):
Idea to avoid `goto`:

```diff
diff --git a/src/quick-lint-js/parse.h b/src/quick-lint-js/parse.h
index 32a64bfe..14ead941 100644
--- a/src/quick-lint-js/parse.h
+++ b/src/quick-lint-js/parse.h
@@ -1450,7 +1450,27 @@ class parser {
             std::optional<identifier> property_name,
             source_code_span property_name_span,
             function_attributes method_attributes) -> void {
-      next:
+        if (async_static && !this->peek().has_leading_newline) {
+          switch (this->peek().type) {
+          case token_type::identifier:
+          case token_type::private_identifier:
+          case token_type::star:
+            // async static method() {}             // Invalid
+            // async static *myAsyncGenerator() {}  // Invalid
+            this->error_reporter_->report(error_async_static_method{
+                .async_static = source_code_span(async_token_begin,
+                                                 property_name_span.end()),
+            });
+            property_name = this->peek().identifier_name();
+            property_name_span = property_name->span();
+            this->skip();
+            break;
+
+          default:
+            break;
+          }
+        }
+
         switch (this->peek().type) {
         // method() { }
         // method { }    // Invalid (missing parameter list).
@@ -1486,17 +1506,6 @@ class parser {
             //   method() {}
             // }
             v.visit_property_declaration(property_name);
-          } else if (async_static) {
-            // async static method() {}             // Invalid
-            // async static *myAsyncGenerator() {}  // Invalid
-            this->error_reporter_->report(error_async_static_method{
-                .async_static = source_code_span(async_token_begin,
-                                                 property_name_span.end()),
-            });
-            property_name = this->peek().identifier_name();
-            property_name_span = property_name->span();
-            this->skip();
-            goto next;
           } else {
             this->error_reporter_->report(error_unexpected_token{
                 .token = property_name_span,
```

erlliam commented:
> Actually, this won't work. What I would like to do is use goto next (the original next, not the one I added), but I don't know how to do that. I guess the original next doesn't exist inside of the lambda. I'm fighting C++.

It turns out what I wanted was just to use the cases from the switch statement where we enter the `parse_and_visit_field_or_method`. No goto needed (especially after your suggestion).

Something sketchy:
I've refactored the cases used (where we enter `parse_and_visit_field...`), perhaps that should land as it's own commit?
Also the test at the could just use parse_and_visit_statement instead of manually creating the spy visitor, etc.

strager commented on test/test-parse-class.cpp:
```diff
@@ -1078,6 +1078,44 @@ TEST(test_parse, class_statement_as_with_statement_body_is_disallowed) {
+    parser p(&code, &v);
+    EXPECT_TRUE(p.parse_and_visit_statement(v));
+    EXPECT_EQ(v.property_declarations[0].name, u8"static");
+  }
```
Should we be checking for `error_async_static_method` here?

strager commented on src/quick-lint-js/parse.h:
```diff
@@ -1549,6 +1576,11 @@ class parser {
       }
+      if (this->peek().type == token_type::kw_static) {
+        // async static method() {}  // Invalid
+        // async static() {}
```
Is `class { async static() {} }` already tested by an existing test? Or should we add a test for it?

erlliam replied:
I believe this test `class_statement_with_keyword_property` handles this.

strager commented on test/test-parse-class.cpp:
```diff
@@ -1078,6 +1078,44 @@ TEST(test_parse, class_statement_as_with_statement_body_is_disallowed) {
+    EXPECT_TRUE(p.parse_and_visit_statement(v));
+    EXPECT_EQ(v.property_declarations[0].name, u8"static");
+  }
+}
```
What about these:
```js
class C {
  async    // V8 treats this as a field declaration, I think.
  static   // V8 treats this as part of the method 'm'.
  m() { await(myPromise); }  // V8 makes a non-async static method called 'm'.
}
```
```js
class C {
  async  // V8 treats this as a field declaration, I think.
  static m() { }  // V8 makes a non-async static method called 'm'.
}
```

strager commented on src/quick-lint-js/parse.h:
```diff
@@ -1441,12 +1441,38 @@ class parser {
+              .async_static =
+                  source_code_span(async_token_begin, property_name_span.end()),
+          });
+          property_name = this->peek().identifier_name();
```
Blocking: This is invalid for `token_type::star`.

strager reviewed (changes requested):
Looks good except for `async static *m() {}`. (Write a test!)

erlliam commented on src/quick-lint-js/parse.h:
```diff
@@ -1540,14 +1569,36 @@ class parser {
+        case token_type::string:
+        case token_type::identifier:
+        case token_type::private_identifier:
+        case token_type::star:
```
This seems wrong (star case)

erlliam commented on src/quick-lint-js/parse.h:
```diff
@@ -1576,12 +1627,8 @@ class parser {
     // #field = initialValue;
     QLJS_CASE_RESERVED_KEYWORD_EXCEPT_FUNCTION:
-    QLJS_CASE_CONTEXTUAL_KEYWORD_EXCEPT_ASYNC_AND_GET_AND_SET_AND_STATIC:
+    QLJS_CASE_CONTEXTUAL_KEYWORD:
```
This change is probably out of scope.

strager commented on test/test-parse-class.cpp:
```diff
@@ -1078,6 +1078,91 @@ TEST(test_parse, class_statement_as_with_statement_body_is_disallowed) {
+                            "visit_property_declaration",       // m
+                            "visit_enter_function_scope",       // m
+                            "visit_enter_function_scope_body",  // m
+                            "visit_variable_use",         // await myPromise
```
Nit: Comment should say just `myPromise`:
```suggestion
                            "visit_variable_use",         // myPromise
```

strager commented on test/test-parse-class.cpp:
```diff
@@ -1078,6 +1078,91 @@ TEST(test_parse, class_statement_as_with_statement_body_is_disallowed) {
+  }
+}
+
+TEST(test_parse, static_method_allows_newline_after_static_keyword) {
+  {
+    spy_visitor v =
+        parse_and_visit_statement(u8"class C { static\n m() { } }"_sv);
+    EXPECT_EQ(v.property_declarations[0].name, u8"m");
+  }
+}
```
Nit: Are `static async m(){}` or `static *m(){}` parsed differently enough to warrant tests?

strager commented on test/test-parse-class.cpp:
```diff
@@ -1078,6 +1078,91 @@ TEST(test_parse, class_statement_as_with_statement_body_is_disallowed) {
+    EXPECT_EQ(v.property_declarations[1].name, u8"m");
+  }
+
+  {
+    spy_visitor v =
+        parse_and_visit_statement(u8"class C { async\n static m() { } }"_sv);
+    EXPECT_EQ(v.property_declarations[0].name, u8"async");
+    EXPECT_EQ(v.property_declarations[1].name, u8"m");
+  }
```
Love it!

strager commented on src/quick-lint-js/parse.h:
```diff
@@ -1540,14 +1569,36 @@ class parser {
     // async f() {}
     case token_type::kw_async:
       last_ident = this->peek().identifier_name();
+      async_token_begin = this->peek().span().begin();
```
Nit:
```suggestion
      async_token_begin = this->peek().begin;
```

strager commented on src/quick-lint-js/parse.h:
```diff
@@ -1540,14 +1569,36 @@ class parser {
-        method_attributes = function_attributes::async_generator;
-        this->skip();
+      if (this->peek().has_leading_newline) {
+        switch (this->peek().type) {
+        QLJS_CASE_KEYWORD:
+        case token_type::left_square:
+        case token_type::number:
+        case token_type::string:
+        case token_type::identifier:
+        case token_type::private_identifier:
+        case token_type::star:
+          v.visit_property_declaration(last_ident);
+          return;
```
Add a comment:
```suggestion
        switch (this->peek().type) {
        // 'async' is a field name:
        // class {
        //   async
        //   method() {}
        // }
        QLJS_CASE_KEYWORD:
        case token_type::left_square:
        case token_type::number:
        case token_type::string:
        case token_type::identifier:
        case token_type::private_identifier:
        case token_type::star:
          v.visit_property_declaration(last_ident);
          return;
```

strager commented on src/quick-lint-js/parse.h:
```diff
@@ -1441,12 +1441,41 @@ class parser {
+            std::optional<identifier> property_name,
+            source_code_span property_name_span,
+            function_attributes method_attributes) -> void {
+      if (async_static && !this->peek().has_leading_newline) {
```
What is the purpose of `!this->peek().has_leading_newline`? It looks like `!this->peek().has_leading_newline` would cause the following examples to parse differently. Am I correct?
```js
class C {
  async static m() {}
}
```
```js
class C {
  async static m
    () {}
}
```

erlliam replied:
I'm not sure what the purpose of this new line check is anymore.

strager commented on src/quick-lint-js/parse.h:
```diff
@@ -1441,12 +1441,41 @@ class parser {
+        case token_type::star:
+          method_attributes = function_attributes::generator;
+          this->skip();
+          [[fallthrough]];
```
Blocking: I think fallthrough after skipping is dangerous. What if `*` is followed by `+`, for example? I think that'd cause a crash in `identifier_name()` on line 1470.

strager commented on src/quick-lint-js/parse.h:
```diff
@@ -1441,12 +1441,41 @@ class parser {
+      if (async_static && !this->peek().has_leading_newline) {
+        switch (this->peek().type) {
+        case token_type::star:
+          method_attributes = function_attributes::generator;
```
Shouldn't this be an async generator?

strager commented on src/quick-lint-js/parse.h:
```diff
@@ -1441,12 +1441,41 @@ class parser {
+          property_name_span = property_name->span();
+          this->skip();
+          break;
+        default:
```
Maybe this deserves a comment?
```suggestion
        // async static() {}
        default:
```

strager commented on src/quick-lint-js/parse.h:
```diff
@@ -1441,12 +1441,41 @@ class parser {
+            function_attributes method_attributes) -> void {
+      if (async_static) {
+        if (this->peek().type == token_type::star) {
+          method_attributes = function_attributes::async_generator;
```
Nit: Comment?
```suggestion
          // async static *m() {}  // Invalid.
          method_attributes = function_attributes::async_generator;
```
